### DISCUSSION TITLE
Feature request: Deployment story of apps from test/staging to prod

### DISCUSSION THREAD

**Discussion by stephanbisser** [pnp/generator-teams#296] [Ideas]
### 💡 Idea

It would be nice to have an easy way of deploying a Teams app which is already deployed on a test/staging tenant to a prod tenant without the need of running through everything again

### Is your feature related to a bug

no

**Comment by appieschot** [comment]
Would that be part of the generator? Or could we reuse some of the cli commands for this?

**Comment by stephanbisser** [comment]
That's exactly what we should discuss 😀

**Comment by appieschot** [comment]
So we have the following commands in the CLI: 

- teams app install
- teams app publish 
- teams app list

Do we miss commands to achieve a redeployment on a different tenant?

**Comment by wictorwilen** [comment]
Could this pipeline be expanded to contain the app publishing as well? https://www.wictorwilen.se/blog/deploying-yo-teams-and-node-apps/

**Comment by stephanbisser** [comment]
Maybe just to clarify my initial idea so we can discuss that:
What would be cool is to have an easy way of deploying a Teams app which I already have deployed to my test tenant for testing while developing it, to my production tenant. The problem I see with a pipeline is, that not everyone uses CI/CD so the CLI would be a natural candidate for that. But it should be as easy as just one command which mainly copies over everything from Dev to Prod as well as deploying the stuff into the org's app store.
@appieschot do you think that the CLI could ease that process?
@wictorwilen do you see pipelines as an optional piece in here or would you say it should be mandatory using a pipeline in here?

**Comment by appieschot** [comment]
@stephanbisser that would fit our vision for the CLI, we could spec out a command that would grab an app and copy it to a different tenant (regardless of dev / test etc). @pnp/cli-for-microsoft-365-maintainers if we agree we could move this issue to the cli repo and spec out the command structure

**Comment by wictorwilen** [comment]
To clarify **deployment** do we mean the actual application (web app, bot) or the Teams package? The latter has support using the `--deploy` flag

**Comment by waldekmastykarz** [comment]
> @stephanbisser that would fit our vision for the CLI, we could spec out a command that would grab an app and copy it to a different tenant (regardless of dev / test etc). @pnp/cli-for-microsoft-365-maintainers if we agree we could move this issue to the cli repo and spec out the command structure

One challenge I can see here is the support for two accounts across the different tenants. We could do this using MSAL, but right now we don't support multiple accounts, so the only option would be to do it in a script with app-only login included.

**Comment by stephanbisser** [comment]
> One challenge I can see here is the support for two accounts across the different tenants. We could do this using MSAL, but right now we don't support multiple accounts, so the only option would be to do it in a script with app-only login included.

That definitely would be something we would need to discuss beforehand what's the best approach to that so it's easy to use and intuitive without the need of login here and there

**Comment by waldekmastykarz** [comment]
If you've got two tenants with different AADs and using different accounts, I don't think there is a way around it. If you can use the same account on both tenants, then we could get away with a single login, but then would still need to add support for selecting the tenant with which you want to communicate.
